File: frontend/src/arkham/types/FlavorText.ts
import * as JsonDecoder from 'ts.data.json';

export type FlavorTextModifier = 'BlueEntry' | 'NestedEntry' | 'RightAligned' | 'PlainText' | 'InvalidEntry' | 'ValidEntry' | 'EntrySplit' | 'ResolutionEntry' | 'CheckpointEntry' | 'InterludeEntry'

export type ImageModifier = 'RemoveImage' | 'SelectImage'

export interface ListItemEntry {
  entry: FlavorTextEntry;
  nested: ListItemEntry[];
}

export type FlavorTextEntry
  = { tag: 'BasicEntry', text : string}
  | { tag: 'I18nEntry', key: string, variables: Record<string, any> }
  | { tag: 'HeaderEntry', key: string }
  | { tag: 'InvalidEntry', text: string }
  | { tag: 'ValidEntry', text: string }
  | { tag: 'ModifyEntry', modifiers: FlavorTextModifier[], entry: FlavorTextEntry }
  | { tag: 'CompositeEntry', entries: FlavorTextEntry[] }
  | { tag: 'ColumnEntry', entries: FlavorTextEntry[] }
  | { tag: 'ListEntry', list: ListItemEntry[] }
  | { tag: 'CardEntry', cardCode: string, imageModifiers: ImageModifier[] }
  | { tag: 'EntrySplit' }

export type FlavorText = {
  title: string | null;
  body: FlavorTextEntry[];
}

export const imageModifierDecoder = JsonDecoder.oneOf<ImageModifier>([
  JsonDecoder.literal('RemoveImage'),
  JsonDecoder.literal('SelectImage'),
], 'ImageModifier');

export const flavorTextModifierDecoder = JsonDecoder.oneOf<FlavorTextModifier>([
  JsonDecoder.literal('BlueEntry'),
  JsonDecoder.literal('NestedEntry'),
  JsonDecoder.literal('ResolutionEntry'),
  JsonDecoder.literal('CheckpointEntry'),
  JsonDecoder.literal('InterludeEntry'),
  JsonDecoder.literal('RightAligned'),
  JsonDecoder.literal('PlainText'),
  JsonDecoder.literal('InvalidEntry'),
  JsonDecoder.literal('ValidEntry'),
], 'FlavorTextModifier');

export const listItemEntryDecoder: JsonDecoder.Decoder<ListItemEntry> = JsonDecoder.object<ListItemEntry>(
  {
    entry: JsonDecoder.lazy(() => flavorTextEntryDecoder),
    nested: JsonDecoder.lazy(() => JsonDecoder.array(listItemEntryDecoder, 'ListItemEntry[]'))
  },
  'ListItemEntry',
);

export const flavorTextEntryDecoder: JsonDecoder.Decoder<FlavorTextEntry> = JsonDecoder.oneOf<FlavorTextEntry>([
  JsonDecoder.object({ tag: JsonDecoder.literal('BasicEntry'), text: JsonDecoder.string() }, 'BasicEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('I18nEntry'), key: JsonDecoder.string(), variables: JsonDecoder.succeed()}, 'I18nEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('HeaderEntry'), key: JsonDecoder.string()}, 'HeaderEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('InvalidEntry'), text: JsonDecoder.string() }, 'InvalidEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('ValidEntry'), text: JsonDecoder.string() }, 'ValidEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('ModifyEntry'), modifiers: JsonDecoder.array(flavorTextModifierDecoder, 'FlavorTextModifier[]'), entry: JsonDecoder.lazy(() => flavorTextEntryDecoder) }, 'ModifyEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('CompositeEntry'), entries: JsonDecoder.lazy(() => JsonDecoder.array(flavorTextEntryDecoder, 'FlavorTextEntry[]')) }, 'CompositeEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('ColumnEntry'), entries: JsonDecoder.lazy(() => JsonDecoder.array(flavorTextEntryDecoder, 'FlavorTextEntry[]')) }, 'CompositeEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('ListEntry'), list: JsonDecoder.array(listItemEntryDecoder, 'ListItemEntry[]') }, 'ListEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('CardEntry'), cardCode: JsonDecoder.string(), imageModifiers: JsonDecoder.array(imageModifierDecoder, 'ImageModifiers[]') }, 'CardEntry'),
  JsonDecoder.object({ tag: JsonDecoder.literal('EntrySplit')}, 'EntrySplit'),
], 'FlavorTextEntry');


export const flavorTextDecoder: JsonDecoder.Decoder<FlavorText> = JsonDecoder.object<FlavorText>(
  {
    title: JsonDecoder.nullable(JsonDecoder.string()),
    body: JsonDecoder.array(flavorTextEntryDecoder, 'string[]')
  },
  'FlavorText',
);

File: backend/arkham-api/library/Arkham/FlavorText.hs
{-# LANGUAGE ImplicitParams #-}
{-# LANGUAGE ImpredicativeTypes #-}
{-# OPTIONS_GHC -Wno-orphans #-}

module Arkham.FlavorText (
  module X,
  flavorText,
  ul,
  li,
  compose,
  h,
  hr,
  p,
  cols,
  img,
  UlItems,
)
where

import Arkham.I18n as X
import Arkham.Text as X

import Arkham.Card.CardCode
import Arkham.Prelude
import Control.Monad.Writer.Strict
import Data.Text qualified as T
import GHC.Records

flavorText :: FlavorTextEntry -> FlavorText
flavorText = FlavorText Nothing . pure

type UlItems = Writer [ListItemEntry] ()

ul :: UlItems -> FlavorTextEntry
ul = ListEntry . execWriter

h :: HasI18n => Scope -> FlavorTextEntry
h t = HeaderEntry (intercalate "." (?scope <> [t]))

p :: HasI18n => Scope -> FlavorTextEntry
p = i18nEntry

compose :: [FlavorTextEntry] -> FlavorTextEntry
compose = CompositeEntry

cols :: [FlavorTextEntry] -> FlavorTextEntry
cols = ColumnEntry

img :: CardCode -> FlavorTextEntry
img = (`CardEntry` [])

hr :: FlavorTextEntry
hr = EntrySplit

li :: HasI18n => Text -> UlItems
li t = tell [ListItemEntry (i18nEntry t) []]

specialize :: (Text -> UlItems) -> Text -> (ListItemEntry -> ListItemEntry) -> UlItems
specialize f t convert = tell $ map convert $ execWriter $ f t

instance HasField "nested" (Text -> UlItems) (String -> UlItems -> UlItems) where
  getField f t items = specialize f (T.pack t) \case
    ListItemEntry entry nested -> ListItemEntry entry (nested <> execWriter items)

instance HasField "valid" (Text -> UlItems) (String -> UlItems) where
  getField f t = specialize f (T.pack t) \case
    ListItemEntry entry nested -> case entry of
      ModifyEntry modifiers inner -> ListItemEntry (ModifyEntry (ValidEntry : modifiers) inner) nested
      _ -> ListItemEntry (ModifyEntry [ValidEntry] entry) nested

instance HasField "invalid" (Text -> UlItems) (String -> UlItems) where
  getField f t = specialize f (T.pack t) \case
    ListItemEntry entry nested -> case entry of
      ModifyEntry modifiers inner -> ListItemEntry (ModifyEntry (InvalidEntry : modifiers) inner) nested
      _ -> ListItemEntry (ModifyEntry [ValidEntry] entry) nested

instance HasField "validate" (Text -> UlItems) (Bool -> String -> UlItems) where
  getField f cond t =
    let modifier = if cond then ValidEntry else InvalidEntry
     in specialize f (T.pack t) \case
          ListItemEntry entry nested -> case entry of
            ModifyEntry modifiers inner -> ListItemEntry (ModifyEntry (modifier : modifiers) inner) nested
            _ -> ListItemEntry (ModifyEntry [modifier] entry) nested

instance HasField "remove" (CardCode -> FlavorTextEntry) (CardCode -> FlavorTextEntry) where
  getField f cardCode = case f cardCode of
    CardEntry _ modifiers -> CardEntry cardCode (RemoveImage : modifiers)
    other -> other

extendModifiers :: FlavorTextModifier -> FlavorTextEntry -> FlavorTextEntry
extendModifiers modifier = \case
  ModifyEntry modifiers inner -> ModifyEntry (modifier : modifiers) inner
  other -> ModifyEntry [modifier] other

instance HasField "blue" (Scope -> FlavorTextEntry) (Scope -> FlavorTextEntry) where
  getField f = extendModifiers BlueEntry . f

instance HasField "right" (Scope -> FlavorTextEntry) (Scope -> FlavorTextEntry) where
  getField f = extendModifiers RightAligned . f

instance HasField "nested" (Scope -> FlavorTextEntry) (Scope -> FlavorTextEntry) where
  getField f = extendModifiers NestedEntry . f

instance HasField "validate" (Scope -> FlavorTextEntry) (Bool -> Scope -> FlavorTextEntry) where
  getField f cond = extendModifiers (if cond then ValidEntry else InvalidEntry) . f
